modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/TransportRankEvalAction.java

```diff
@@ -57,56 +49,75 @@
+            String[] indices = new String[querySpecification.getIndices().size()];
+            querySpecification.getIndices().toArray(indices);
             SearchRequest templatedRequest = new SearchRequest(indices, specRequest);
-            String[] types = new String[spec.getTypes().size()];
-            spec.getTypes().toArray(types);
+            String[] types = new String[querySpecification.getTypes().size()];
+            querySpecification.getTypes().toArray(types);
             templatedRequest.types(types);
+            client.search(templatedRequest, searchListener);
+        }
+    }
+
+    public static class RankEvalActionListener implements ActionListener<SearchResponse> {
 
-            TransportSearchAction transportSearchAction = new TransportSearchAction(settings, threadPool, searchPhaseController,
-                    transportService, searchTransportService, clusterService, actionFilters, indexNameExpressionResolver);
-            ActionFuture<SearchResponse> searchResponse = transportSearchAction.execute(templatedRequest);
-            SearchHits hits = searchResponse.actionGet().getHits();
+        private ActionListener<RankEvalResponse> listener;
+        private QuerySpec specification;
+        private Map<String, EvalQueryQuality> partialResults;
+        private RankEvalSpec task;
+        private Map<String, Collection<String>> unknownDocs;
+        private AtomicInteger responseCounter;
 
-            EvalQueryQuality queryQuality = metric.evaluate(hits.getHits(), spec.getRatedDocs());
-            partialResults.addElement(queryQuality);
-            unknownDocs.put(spec.getSpecId(), queryQuality.getUnknownDocs());
+        public RankEvalActionListener(ActionListener<RankEvalResponse> listener, RankEvalSpec task, QuerySpec specification,
+                Map<String, EvalQueryQuality> partialResults, Map<String, Collection<String>> unknownDocs, AtomicInteger responseCounter) {
+            this.listener = listener;
+            this.task = task;
+            this.specification = specification;
+            this.partialResults = partialResults;
+            this.unknownDocs = unknownDocs;
+            this.responseCounter = responseCounter;
         }
 
-        // TODO add other statistics like micro/macro avg?
-        RankEvalResponse response = new RankEvalResponse(qualityTask.getTaskId(), metric.combine(partialResults), unknownDocs);
-        listener.onResponse(response);
+        @Override
+        public void onResponse(SearchResponse searchResponse) {
+            SearchHits hits = searchResponse.getHits();
+            EvalQueryQuality queryQuality = task.getEvaluator().evaluate(hits.getHits(), specification.getRatedDocs());
+            partialResults.put(specification.getSpecId(), queryQuality);
+            unknownDocs.put(specification.getSpecId(), queryQuality.getUnknownDocs());
+
+            if (responseCounter.decrementAndGet() == 0) {
```

Comment:
Not sure if that makes any sense at all, but somehow my inclination would be to write counter < 1 instead of counter == 0 here.